subprojects/gst-plugins-base/gst-libs/gst/gl/egl/gsteglcaps.c

```diff
@@ -0,0 +1,287 @@
+                          EGLDisplay dpy,
+                          EGLint format,
+                          EGLint max_modifiers,
+                          EGLuint64KHR *modifiers,
+                          EGLBoolean *external_only,
+                          EGLint *num_modifiers);
+  gst_eglQueryDmaBufModifiersEXT =
+      gst_gl_context_get_proc_address (context, "eglQueryDmaBufModifiersEXT");
+  if (!gst_eglQueryDmaBufModifiersEXT)
+    return FALSE;
+
+  gl_dpy_egl = gst_gl_display_egl_from_gl_display (context->display);
+  if (!gl_dpy_egl) {
+    GST_WARNING_OBJECT (context,
+        "Failed to retrieve GstGLDisplayEGL from %" GST_PTR_FORMAT,
+        context->display);
+    return FALSE;
+  }
+  egl_dpy =
+      (EGLDisplay) gst_gl_display_get_handle (GST_GL_DISPLAY (gl_dpy_egl));
+  gst_object_unref (gl_dpy_egl);
+
+  ret = gst_eglQueryDmaBufModifiersEXT (egl_dpy,
+      fourcc, 0, NULL, NULL, &num_mods);
+  if (!ret || num_mods == 0) {
+      return FALSE;
+  }
+
+  mods = g_new (EGLuint64KHR, num_mods);
+  ret = gst_eglQueryDmaBufModifiersEXT (egl_dpy,
+      fourcc, num_mods, mods, NULL, &num_mods);
+  if (!ret || num_mods == 0) {
+    g_free (mods);
+    return FALSE;
+  }
+
+  g_value_init (&gmod, G_TYPE_UINT64);
+  for (int i = 0; i < num_mods; ++i) {
+    g_value_set_uint64 (&gmod, mods[i]);
+    gst_value_list_append_value (modifiers, &gmod);
+  }
+  g_free (mods);
+
+  return TRUE;
+}
+
+static gboolean
+dma_formats_to_modifier_formats (const GValue *dma_fmts,
+    GValue *dma_mod_fmts)
+{
+  gint num_fmts = 0;
+  GstGLContext *context;
+  GstGLDisplay *display;
+
+  display = GST_GL_DISPLAY (gst_gl_display_egl_new ());
+  context = gst_gl_context_new (display);
+  gst_gl_context_create (context, 0, NULL);
+
+  num_fmts = gst_value_list_get_size (dma_fmts);
+
+  for (gint i = 0; i < num_fmts; i++) {
+    const GValue *gfmt;
+    const gchar *fmt_str;
+    GstVideoFormat fmt;
+    guint32 fcc;
+    GValue mods = G_VALUE_INIT;
+
+    gfmt = gst_value_list_get_value (dma_fmts, i);
+    fmt_str = g_value_get_string (gfmt);
+    fmt = gst_video_format_from_string (fmt_str);
+    fcc = gst_video_format_to_fourcc (fmt);
+
+    g_value_init (&mods, GST_TYPE_LIST);
+    if (!gst_gl_get_modifiers (context, fcc, &mods)){
+      fmts_append (dma_mod_fmts, fmt_str);
+    }
+
+    for (gint j = 0; j < gst_value_list_get_size (&mods); j++) {
+      const GValue *gmod = gst_value_list_get_value (&mods, j);
+      guint64 mod = g_value_get_uint64 (gmod);
+      GST_DEBUG ("Got mofidier: %s:0x%016lx", fmt_str, mod);
+
+      if (mod == DRM_FORMAT_MOD_INVALID)
+        continue;
+
+      if (mod == DRM_FORMAT_MOD_LINEAR)
+        fmts_append (dma_mod_fmts, fmt_str);
+      else
+        mod_fmts_append (dma_mod_fmts, fmt_str, mod);
+    }
+    g_value_unset (&mods);
+  }
+
+  gst_object_unref (context);
+  gst_object_unref (display);
+
+  return TRUE;
+}
+
+GstCaps *
+gst_egl_caps_embed_modifiers (GstCaps * caps, GstGLContext * context)
```

Comment:
I notice that you really do not use this "context" here and always create a new GL context to detect.
I think you should use this context first. And no it is NULL, create a new one.